api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,400 @@ package v1alpha1
+
+	// FilterType marks the Name or Pattern field
+	// as an included or excluded set of domains.
+	//
+	// In the event of contradicting domain options,
+	// preference is given to excluded domains.
+	//
+	// This field accepts the following values:
+	//
+	//  "Include": Include the domain set specified
+	//  by name or pattern.
+	//
+	//  "Exclude": Exclude the domain set specified
+	//  by name or pattern.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	FilterType ExternalDNSFilterType `json:"filterType"`
 }
 
-//+kubebuilder:object:root=true
-//+kubebuilder:subresource:status
+// ExternalDNSDomainUnion describes optional fields of an External domain
+// that should be captured.
+// +union
+type ExternalDNSDomainUnion struct {
+	// MatchType specifies the type of match to be performed
+	// by ExternalDNS when determining whether or not to publish DNS
+	// records for a given source resource based on the resource's
+	// requested hostname.
+	//
+	// This field accepts the following values:
+	//
+	//  "Exact": Explicitly match the full domain string
+	//   specified via the name field, including any subdomains
+	//   of the given name.
+	//
+	//  "Pattern": Match potential domains against
+	//  the provided regular expression pattern string.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	MatchType DomainMatchType `json:"matchType"`
 
-// ExternalDNS is the Schema for the externaldns API
-type ExternalDNS struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Name is a string representing a single domain
+	// value.
+	//
+	// e.g. my-app.my-cluster-domain.com
+	//
+	// +optional
+	Name *string `json:"names,omitempty"`
+
+	// Pattern is a regular expression used to
+	// match a set of domains. Any provided
+	// regular expressions should follow the syntax
+	// used by the go regexp package (RE2).
+	// See https://golang.org/pkg/regexp/ for more information.
+	//
+	// +optional
+	Pattern *string `json:"pattern,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=Exact;Pattern
+type DomainMatchType string
+
+const (
+	DomainMatchTypeExact DomainMatchType = "Exact"
+	DomainMatchTypeRegex DomainMatchType = "Pattern"
+)
+
+// +kubebuilder:validation:Enum=Include;Exclude
+type ExternalDNSFilterType string
+
+const (
+	FilterTypeInclude ExternalDNSFilterType = "Include"
+	FilterTypeExclude ExternalDNSFilterType = "Exclude"
+)
+
+// ExternalDNSProvider specifies configuration
+// options for the desired ExternalDNS DNS provider.
+type ExternalDNSProvider struct {
+	ExternalDNSProviderUnion `json:",inline"`
+
+	// Credentials is a reference to a secret containing
+	// the relevant credentials for the given provider, to be
+	// used by an ExternalDNS controller for creating, updating,
+	// and viewing DNS records.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Credentials NamespacedSecretReference `json:"credentials"`
```

Comment:
The credentials format seem to vary from provider to provider. InfoBlox uses username/password, Azure and BlueCat some json file config, AWS (in case kube2iam is used) needs IAM role name, etc.
I can think of 2 options here:
- elaborate the comment with expected formats for different providers
- add dedicated fields for each supported provider

Reply:
I believe the Bluecat provider is moving away from a json config file for storing credentials and other information. Let me double check how the other providers handle credentials and update this accordingly.

Reply:
I've gone ahead and added fields for each providers specific configuration options.

Reply:
> I believe the Bluecat provider is moving away from a json config file for storing credentials and other information. Let me double check how the other providers handle credentials and update this accordingly.

If the format is not owned by us, how do you plan to plan to handle changes over time and smooth upgrades?

Reply:
> If the format is not owned by us, how do you plan to plan to handle changes over time and smooth upgrades?

Good question. We will have to work with upstream around issues like this (perhaps we could maintain backwards compatibility upstream to help with upgrades).